### DISCUSSION TITLE
Placing nodes inside of other nodes?

### DISCUSSION THREAD

**Discussion by Ginxss** [eclipse-glsp/glsp#692] [Q&A] [answered]
Hi,
I'm making a UML Package Diagram for a Uni-Project with GLSP.

I used the Java-EMF-Theia project template (https://github.com/eclipse-glsp/glsp-examples/tree/master/project-templates/java-emf-theia) and used the UML2 metamodel as a source model instead of the tasklist model from the example.
I can place UML packages at a root level just like the tasks in the example.

You can probably already guess my question. I now want to be able to place packages inside of other packages (or at least on top of them).
Currently, when I select the package creation tool and hover my mouse over another package, it looks like this:
![img](https://user-images.githubusercontent.com/13721331/176166408-74ad9e98-e580-422a-b3ff-238f6270ae60.png)

How would I "unlock" the behaviour of placing nodes on top of or inside of other nodes?

**Comment by CamilleLetavernier** [comment, marked as answer]
Hi @Ginxss,

You can check how we did it for "Category" nodes in the Example workflow. You need:

- A Compartment inside of your package, below the label, that will receive child nodes (e.g. Classes or Packages)
- A "Freeform" layout on this Compartment (Essentially no layout at all; only manual positioning/resizing)
- And you need to update the hints on the server (In your DiagramConfiguration), to indicate which element(s) can be created on/in a Package:

```java
ShapeTypeHint catHint = new ShapeTypeHint(CATEGORY, true, true, true, true);
      catHint.setContainableElementTypeIds(
         Arrays.asList(DECISION_NODE, MERGE_NODE, FORK_NODE, JOIN_NODE, AUTOMATED_TASK, MANUAL_TASK, CATEGORY));
```

The model for Categories looks like this:

- Category [layout=vbox]
  - Header [label]
  - StructCompartment [compartment, layout=freeform, vGrab=true, hGrab=true]
    - ChildTask...
    - ChildTask...
    - ChildCategory...
    - ...
